{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "import math"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The init method initializes all parameters to zero. \n",
    "\n",
    "### If any parameter needs to have a default value, you can set it here. \n",
    "\n",
    "### All penalty parameters are set to a default value of 10"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(object):\n",
    "    \n",
    "    def __init__ (self, name):\n",
    "        #parameters are based on the biology handout\n",
    "        '''parameters for the length criterion'''\n",
    "        self.max_length = 22\n",
    "        self.min_length = 18\n",
    "        self.penalty_length = 10\n",
    "        \n",
    "        '''parameters for the temperature difference criterion'''\n",
    "        self.max_tdiff = 1\n",
    "        self.min_tdiff = 0\n",
    "        self.penalty_tdiff = 10\n",
    "        \n",
    "        '''parameters for the cg content criterion'''\n",
    "        self.max_cg = 0.6\n",
    "        self.min_cg = 0.4\n",
    "        self.penalty_cg = 10\n",
    "        \n",
    "        '''parameters for the annealing temperature criterion'''\n",
    "        self.max_temp = self.max_length*(0.6)*(4) + self.max_length*(0.4)*(2) #70.4 degrees celcius\n",
    "        self.min_temp = self.min_length*(0.4)*(4) + self.min_length*(0.6)*(2) #50.4 degrees celcius\n",
    "        self.penalty_temp = 10\n",
    "        \n",
    "        '''parameters for the run criterion'''\n",
    "        self.run_threshold = 4\n",
    "        self.penalty_runs = 10\n",
    "        \n",
    "        '''parameters for the repeat criterion'''\n",
    "        self.repeat_threshold = 1\n",
    "        self.penalty_repeats = 10\n",
    "        \n",
    "        '''parameters for the specificity criterion'''\n",
    "        self.penalty_specificity = 10 \n",
    "        \n",
    "        '''locations where the forward primer should be chosen from'''\n",
    "        self.fp_start = 301\n",
    "        self.fp_end = 600\n",
    "        \n",
    "        '''locations where the reverse primer should be chosen from'''\n",
    "        self.rp_start = 601\n",
    "        self.rp_end = 901\n",
    "        \n",
    "        ''' parameters for the simulated annealing portion'''\n",
    "        self.initial_temperature = 70.4 #this is self.max_temp aka max annealing temperature\n",
    "        self.stopping_temperature = 0.01\n",
    "        self.drop_fraction = 0.999\n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 2 \n",
    "This method takes in a string variable containing the raw gene sequence and cleans it up to leave only the alphabet characters a,t,c,g. The result is then stored in the instance variable dna_sequence to be used in other methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "attgggttgccgt\n"
     ]
    }
   ],
   "source": [
    "#removes all non-alphabet characters from the selected dna sequence\n",
    "class PrimerDesign(PrimerDesign): \n",
    "\n",
    "    def set_dna_sequence(self, dna_sequence):\n",
    "        if type(dna_sequence) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        elif dna_sequence == '':\n",
    "            return 0\n",
    "        #changes the string into a list containing all the elements, and recombines it into a string with all non-alphabet characters excluded\n",
    "        else:\n",
    "            dna_sequence = ''.join([i for i in dna_sequence if i.isalpha()]).strip(\"\\n\")\n",
    "            self.dna_sequence = dna_sequence\n",
    "            return dna_sequence\n",
    "#Test\n",
    "a = PrimerDesign(\"BIODW\")\n",
    "c = \"attggg 235 ttgccgt\" \n",
    "b = a.set_dna_sequence(c)\n",
    "print(b)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 3\n",
    "This method generates a forward primer or reverse primer at random with a specified length of 20bp long. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 203,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(PrimerDesign):\n",
    "    \n",
    "    def func_select_random(self, sqtype='forward', length = 20 ):\n",
    "        if self.dna_sequence == 0:\n",
    "            return 'ERROR, dna sequence is not a string'\n",
    "        #ensure length is positive\n",
    "        if length < 0:\n",
    "            return ('Length has to be a positive number!')\n",
    "        \n",
    "        '''the length has to be a positive number'''\n",
    "        \n",
    "        if(sqtype == 'forward'):\n",
    "            start_limit = self.fp_start \n",
    "            end_limit = self.fp_end \n",
    "            #initiating starting position of the forward primer randomly\n",
    "            startt = random.randint(start_limit, end_limit - length +1)\n",
    "            #find the end position of the forward primer\n",
    "            endd = startt + length\n",
    "            #slice the dna sequence to obtain the forward primer with correct length\n",
    "            fprimer = self.dna_sequence[startt:endd+1]\n",
    "            return fprimer\n",
    "            \n",
    "        elif(sqtype == 'reverse'):\n",
    "        \n",
    "            start_limit = self.rp_start \n",
    "            end_limit = self.rp_end\n",
    "            #initiating starting position of the reverse primer randomly\n",
    "            startt = random.randint(start_limit, end_limit-length+1)\n",
    "            #find the end position of the reverse primer\n",
    "            endd = startt + length\n",
    "            #slice the dna sequence to obtain the complement of the reverse primer with correct length\n",
    "            rprimer = self.dna_sequence[startt:endd+1]\n",
    "            #we do not complement it since it which will yield the same values for ALL cost function as the reverse primer,hence we will use the complement for the subsequent functions.\n",
    "            return rprimer\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 4\n",
    "The first three methods determine the length of the sequence, the fraction of 'c's and 'g's in the string of dna sequence, and the annealing temperature. Followed by a method that shows the number of runs throughout the dna sequence; a run is indicated by a base that consecutively appears for more than the run threshold, which is 4 times. The last method determines the number of repeats of two distict bases that appear consecutively. \n",
    "\n",
    "If the input is not a string, the functions will return an error message. If the input is an empty string, the functions will return a 0 value. \n",
    "\n",
    "There are several test cases from eDimension below the last method.\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 204,
   "metadata": {},
   "outputs": [],
   "source": [
    "class PrimerDesign(PrimerDesign): \n",
    "            \n",
    "    def func_length(self, sq):\n",
    "        if type(sq) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        else:\n",
    "            return len(sq)\n",
    "    \n",
    "    def func_cg_fraction(self, sq):\n",
    "        if type(sq) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        elif sq ==\"\":\n",
    "            return 0\n",
    "        else:\n",
    "            dic = {'a' :0, 'c' : 0, 'g' : 0, 't' : 0}\n",
    "            for alphabet in sq:\n",
    "                if alphabet in dic:\n",
    "                    dic[alphabet]+=1\n",
    "            cgfrac = (dic[\"c\"] + dic[\"g\"])/sum(dic.values())\n",
    "            return cgfrac\n",
    "    \n",
    "    def func_temperature(self,sq):\n",
    "        if type(sq) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        else:\n",
    "            dic = {'a' :0, 'c' : 0, 'g' : 0, 't' : 0}\n",
    "            for alphabet in sq:\n",
    "                if alphabet in dic:\n",
    "                    dic[alphabet]+=1\n",
    "            Tsq = 4*(dic['g']+dic['c']) + 2*(dic['a']+dic['t'])\n",
    "            return Tsq\n",
    "            "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class PrimerDesign(PrimerDesign):\n",
    "\n",
    "    def func_count_runs(self,sq):\n",
    "        if type(sq) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        elif sq == '':\n",
    "            return 0\n",
    "        else:\n",
    "            first = sq[0]\n",
    "            #a is the count of the number of times a base appears consecutively\n",
    "            a = 0\n",
    "            lst = []\n",
    "            for alphabet in sq:\n",
    "                if alphabet == first:\n",
    "                    #if the same base appears again, add on to the count \"a\"\n",
    "                    a += 1\n",
    "                else: \n",
    "                    #adds on to the list the number of consecutive appearance once base stops appearing consecutively\n",
    "                    lst.append(a)\n",
    "                    #restarts the count\n",
    "                    a = 0\n",
    "                    first = alphabet\n",
    "                    a+=1\n",
    "            lst.append(a)\n",
    "            counter = 0\n",
    "            for i in lst:\n",
    "                #if base appears consecutively four times, count for runs increase by one\n",
    "                if i > self.run_threshold:\n",
    "                    counter+=1\n",
    "            return counter\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "24\n",
      "0.20833333333333334\n",
      "58\n",
      "2\n",
      "0\n",
      "21\n",
      "0.0\n",
      "42\n",
      "0\n",
      "12\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "ERROR, dna sequence is not a string.\n",
      "ERROR, dna sequence is not a string.\n",
      "ERROR, dna sequence is not a string.\n",
      "ERROR, dna sequence is not a string.\n",
      "ERROR, dna sequence is not a string.\n",
      "11838\n",
      "0.3966067864271457\n",
      "27988\n",
      "36\n",
      "309\n"
     ]
    }
   ],
   "source": [
    "\n",
    "class PrimerDesign(PrimerDesign):\n",
    "    def func_count_repeats(self,sq):\n",
    "        if type(sq) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        \n",
    "        elif sq == '':\n",
    "            return 0\n",
    "        \n",
    "        else:\n",
    "            di_repeats = ['at','ac','ag','ca','ct','cg','ga','gt','gc','ta','tc','tg']\n",
    "            lst = list(sq)\n",
    "            anss = 0\n",
    "            for i in range(len(lst)-3):\n",
    "                #if two or more of the same bases appear consecutively, it is NOT counted as a repeat\n",
    "                if lst[i] == lst[i+1]:\n",
    "                    pass\n",
    "                #if two bases appear consecutively, add on to the count \"anss\"\n",
    "                elif lst[i] == lst[i+2] and lst[i+1] == lst[i+3]:\n",
    "                    anss+=1\n",
    "            return anss\n",
    "\n",
    "#TEST CASE 1   \n",
    "a = PrimerDesign(\"BIODW\")\n",
    "b = 'cattaaaaatacgaaaaaagtcat'\n",
    "print(a.func_length(b))\n",
    "print(a.func_cg_fraction(b))\n",
    "print(a.func_temperature(b))\n",
    "print(a.func_count_runs(b))\n",
    "print(a.func_count_repeats(b))\n",
    "\n",
    "#TEST CASE 2 - this test ensures that func_cg_fraction runs properly (no \"g\"s and \"c\"s in dna sequence) and that runs and repeats are computed accurately\n",
    "b = 'atatatatattttatatataa'\n",
    "print(a.func_length(b))\n",
    "print(a.func_cg_fraction(b))\n",
    "print(a.func_temperature(b))\n",
    "print(a.func_count_runs(b))\n",
    "print(a.func_count_repeats(b))\n",
    "\n",
    "#TEST CASE 3 - this test ensures that empty string returns 0 for all functions\n",
    "b = ''\n",
    "print(a.func_length(b))\n",
    "print(a.func_cg_fraction(b))\n",
    "print(a.func_temperature(b))\n",
    "print(a.func_count_runs(b))\n",
    "print(a.func_count_repeats(b))\n",
    "\n",
    "#TEST CASE 4 - this test ensures that only strings should be accepted as input, otherwise returns an error message\n",
    "b = 1233\n",
    "print(a.func_length(b))\n",
    "print(a.func_cg_fraction(b))\n",
    "print(a.func_temperature(b))\n",
    "print(a.func_count_runs(b))\n",
    "print(a.func_count_repeats(b))\n",
    "\n",
    "#TEST CASE 5 - this test made sure that real and long dna sequences can be computed by the methods as well\n",
    "b = dna_sequence\n",
    "print(a.func_length(b))\n",
    "print(a.func_cg_fraction(b))\n",
    "print(a.func_temperature(b))\n",
    "print(a.func_count_runs(b))\n",
    "print(a.func_count_repeats(b))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 5\n",
    "Each of these class methods calculates the cost of each previous criteria of the dna sequence, e.g. length, runs, repeats, etc. If the cost function for a particular criterion returns zero, the criterion is met. \n",
    "\n",
    "Several test cases from eDimension are displayed below.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 207,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20\n",
      "0\n",
      "1.9166666666666667\n",
      "20\n",
      "0\n",
      "0\n",
      "\n",
      "\n",
      "0\n",
      "83.99999999999999\n",
      "4.0\n",
      "0\n",
      "120\n",
      "0\n",
      "\n",
      "\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "0\n",
      "\n",
      "\n",
      "ERROR, dna sequence is not a string.\n",
      "ERROR, dna sequence is not a string.\n",
      "ERROR, dna sequence is not a string.\n",
      "ERROR, dna sequence is not a string.\n",
      "ERROR, dna sequence is not a string.\n",
      "ERROR, dna sequence is not a string.\n",
      "\n",
      "\n",
      "118160\n",
      "279176.0\n",
      "0.033932135728543256\n",
      "360\n",
      "3090\n",
      "0\n",
      "\n",
      "\n",
      "10\n"
     ]
    }
   ],
   "source": [
    "class PrimerDesign(PrimerDesign):\n",
    "    \n",
    "    def cost_length(self, sq):\n",
    "        if type(sq) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        \n",
    "        elif sq == '':\n",
    "            return 0\n",
    "        \n",
    "        else:\n",
    "            '''This is given to you as an example '''\n",
    "            sq_len = len(sq)\n",
    "            if(sq_len > self.max_length):\n",
    "                return (sq_len - self.max_length)*self.penalty_length\n",
    "            elif(sq_len < self.min_length):\n",
    "                return (self.min_length - sq_len)*self.penalty_length \n",
    "            else:\n",
    "                return 0\n",
    "    \n",
    "    def cost_temperature(self,sq):\n",
    "        if type(sq) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        \n",
    "        elif sq == '':\n",
    "            return 0\n",
    "        \n",
    "        else:\n",
    "            Tsq = self.func_temperature(sq)\n",
    "            if Tsq > self.max_temp:\n",
    "                return self.penalty_temp*(Tsq - self.max_temp)\n",
    "            elif self.min_temp <= Tsq <= self.max_temp:\n",
    "                return 0\n",
    "            elif Tsq < self.min_temp:\n",
    "                return self.penalty_temp*(self.min_temp - Tsq)\n",
    "        \n",
    "    def cost_cgcontent(self,sq):\n",
    "        if type(sq) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        \n",
    "        elif sq == '':\n",
    "            return 0\n",
    "        \n",
    "        else:\n",
    "            c = self.penalty_cg\n",
    "            cgfrac = self.func_cg_fraction(sq)\n",
    "\n",
    "            if cgfrac > self.max_cg:\n",
    "                return c*(cgfrac - self.max_cg)\n",
    "            elif self.min_cg <= cgfrac <= self.max_cg:\n",
    "                return 0\n",
    "            elif cgfrac < self.min_cg:\n",
    "                return c*(self.min_cg - cgfrac)\n",
    "        \n",
    "    def cost_temperature_difference(self, fp, rp):\n",
    "        if type(fp) != str or type(rp) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        \n",
    "        elif fp == '' or rp ==\"\":\n",
    "            return 0\n",
    "        \n",
    "        else:\n",
    "            dic = {'a' :0, 'c' : 0, 'g' : 0, 't' : 0}\n",
    "            dc = {'a' :0, 'c' : 0, 'g' : 0, 't' : 0}\n",
    "            for alphabet in fp:\n",
    "                if alphabet in dic:\n",
    "                    dic[alphabet]+=1 \n",
    "            for alphabet in rp:\n",
    "                if alphabet in dc:\n",
    "                    dc[alphabet]+=1 \n",
    "            Tapr = 4*(dic['g']+dic['c']) + 2*(dic['a']+dic['t'])\n",
    "            Tarp = 4*(dc['g']+dc['c']) + 2*(dc['a']+dc['t'])\n",
    "\n",
    "            deltaT = abs(Tapr - Tarp) \n",
    "            if deltaT > self.max_tdiff:\n",
    "                return self.penalty_tdiff*(deltaT-self.max_tdiff)\n",
    "            else:\n",
    "                return 0\n",
    "    \n",
    "    def cost_specificity(self, sq): \n",
    "        if type(sq) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        \n",
    "        elif sq == '':\n",
    "            return 0\n",
    "        \n",
    "        else:\n",
    "\n",
    "            countdna = self.dna_sequence.count(sq)\n",
    "            if countdna > 0:\n",
    "                return (countdna-1)*self.penalty_specificity\n",
    "            else:\n",
    "                return 0\n",
    "    \n",
    "    \n",
    "    def cost_runs(self, sq):\n",
    "        if type(sq) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        \n",
    "        elif sq == '':\n",
    "            return 0\n",
    "        \n",
    "        else:\n",
    "            b = self.penalty_runs\n",
    "            count = self.func_count_runs(sq)\n",
    "            return b*count\n",
    "    \n",
    "    def cost_repeats(self,sq):\n",
    "        if type(sq) != str:\n",
    "            return \"ERROR, dna sequence is not a string.\"\n",
    "        \n",
    "        elif sq == '':\n",
    "            return 0\n",
    "        \n",
    "        else:\n",
    "            r = self.penalty_repeats\n",
    "            if r>0:\n",
    "                anss = self.func_count_repeats(sq)\n",
    "            else:\n",
    "                anss = 0\n",
    "\n",
    "            return r*anss\n",
    "        \n",
    "#Test cases 1-3 are from eDimension to test all the cost functions are working properly\n",
    "\n",
    "#TEST CASE 1\n",
    "a = PrimerDesign(\"BIODW\")\n",
    "b = 'cattaaaaatacgaaaaaagtcat'\n",
    "a.set_dna_sequence(b)\n",
    "print(a.cost_length(b))\n",
    "print(a.cost_temperature(b))\n",
    "print(a.cost_cgcontent(b))\n",
    "print(a.cost_runs(b))\n",
    "print(a.cost_repeats(b))\n",
    "print(a.cost_specificity(b))\n",
    "print(\"\\n\")\n",
    "\n",
    "#TEST CASE 2\n",
    "b = 'atatatatattttatatataa'\n",
    "a.set_dna_sequence(b)\n",
    "print(a.cost_length(b))\n",
    "print(a.cost_temperature(b))\n",
    "print(a.cost_cgcontent(b))\n",
    "print(a.cost_runs(b))\n",
    "print(a.cost_repeats(b))\n",
    "print(a.cost_specificity(b))\n",
    "print(\"\\n\")\n",
    "\n",
    "#TEST CASE 3 - ensures empty strings have no cost functions.\n",
    "b = ''\n",
    "a.set_dna_sequence(b)\n",
    "print(a.cost_length(b))\n",
    "print(a.cost_temperature(b))\n",
    "print(a.cost_cgcontent(b))\n",
    "print(a.cost_runs(b))\n",
    "print(a.cost_repeats(b))\n",
    "print(a.cost_specificity(b))\n",
    "print(\"\\n\")\n",
    "\n",
    "#TEST CASE 4 - this test ensures that only strings should be accepted as input, otherwise returns an error message.\n",
    "b = 12345\n",
    "a.set_dna_sequence(b)\n",
    "print(a.cost_length(b))\n",
    "print(a.cost_temperature(b))\n",
    "print(a.cost_cgcontent(b))\n",
    "print(a.cost_runs(b))\n",
    "print(a.cost_repeats(b))\n",
    "print(a.cost_specificity(b))\n",
    "print(\"\\n\")\n",
    "\n",
    "#TEST CASE 5 - this test ensures cost functions can run properly on real and long dna sequences.\n",
    "b = dna_sequence\n",
    "a.set_dna_sequence(b)\n",
    "print(a.cost_length(b))\n",
    "print(a.cost_temperature(b))\n",
    "print(a.cost_cgcontent(b))\n",
    "print(a.cost_runs(b))\n",
    "print(a.cost_repeats(b))\n",
    "print(a.cost_specificity(b))\n",
    "print(\"\\n\")\n",
    "\n",
    "#TEST CASE 6 - this test ensures that the cost function for cost_temperature_difference works\n",
    "a = PrimerDesign(\"BIODW\")\n",
    "c = dna_sequence\n",
    "b = a.set_dna_sequence(c)\n",
    "fp = a.func_select_random(sqtype='forward', length =20)\n",
    "rp = a.func_select_random(sqtype='reverse', length =20)\n",
    "print(a.cost_temperature_difference(fp, rp))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 6\n",
    "This is a method to determine the objective function, which is the sum of all the cost functions described in the previous section. It gives you a score that tells you how close these primers are to the optimal primer. A score of zero tells you that your primers selected have met all the criteria. When the score is close to zero, you know that the primers that you have selected is close to the optimal choice. \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 208,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "105.36842105263156\n",
      "\n",
      "\n",
      "50\n",
      "\n",
      "\n",
      "10.190476190476192\n",
      "\n",
      "\n",
      "40\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "class PrimerDesign(PrimerDesign):\n",
    "    \n",
    "    def cost_objective_function(self, fp, rp):\n",
    "        '''complete the calculation of the cost'''\n",
    "        cost = 0 \n",
    "        cost+=self.cost_length(fp)\n",
    "        cost+=self.cost_length(rp)\n",
    "        cost+=self.cost_temperature(fp)\n",
    "        cost+=self.cost_temperature(rp)\n",
    "        cost+=self.cost_cgcontent(fp)\n",
    "        cost+=self.cost_cgcontent(rp)\n",
    "        cost+=self.cost_specificity(fp)\n",
    "        cost+=self.cost_specificity(rp)\n",
    "        cost+=self.cost_runs(fp)\n",
    "        cost+=self.cost_runs(rp)\n",
    "        cost+=self.cost_repeats(fp)\n",
    "        cost+=self.cost_repeats(rp)\n",
    "        cost+=self.cost_temperature_difference(fp,rp)\n",
    "        \n",
    "        return cost \n",
    "\n",
    "#Test 1-4 ensures that all the previous tests run smoothly and checks whether length of primers affect the cost objective function\n",
    "#Test based on dna_sequence\n",
    "a = PrimerDesign(\"BIODW\")\n",
    "c = dna_sequence\n",
    "b = a.set_dna_sequence(c)\n",
    "fp = a.func_select_random(sqtype='forward', length =18)\n",
    "rp = a.func_select_random(sqtype='reverse', length =18)\n",
    "\n",
    "print(a.cost_objective_function(fp,rp))\n",
    "print(\"\\n\")\n",
    "\n",
    "#Test based on dna_sequence(2)\n",
    "a = PrimerDesign(\"BIODW\")\n",
    "c = dna_sequence\n",
    "b = a.set_dna_sequence(c)\n",
    "fp = a.func_select_random(sqtype='forward', length =19)\n",
    "rp = a.func_select_random(sqtype='reverse', length =19)\n",
    "\n",
    "print(a.cost_objective_function(fp,rp))\n",
    "print(\"\\n\")\n",
    "\n",
    "#Test based on dna_sequence(3)\n",
    "a = PrimerDesign(\"BIODW\")\n",
    "c = dna_sequence\n",
    "b = a.set_dna_sequence(c)\n",
    "fp = a.func_select_random(sqtype='forward', length =20)\n",
    "rp = a.func_select_random(sqtype='reverse', length =20)\n",
    "\n",
    "print(a.cost_objective_function(fp,rp))\n",
    "print(\"\\n\")\n",
    "\n",
    "#Test based on dna_sequence(4)\n",
    "a = PrimerDesign(\"BIODW\")\n",
    "c = dna_sequence\n",
    "b = a.set_dna_sequence(c)\n",
    "fp = a.func_select_random(sqtype='forward', length =21)\n",
    "rp = a.func_select_random(sqtype='reverse', length =21)\n",
    "\n",
    "print(a.cost_objective_function(fp,rp))\n",
    "print(\"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 7\n",
    "This instance method displays all the criteria and corresponding cost function score for each of the forward primer and reverse primer in a table. The table provides useful information to clearly see which criterion is met or not met for any pair of primers, e.g. if the cost function for a particular criterion returns zero, the criterion is met. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===Forward Primer=== atctcactaaaagaatagata\n",
      "Criterion                       Cost function score\n",
      "---------------------------------------------------\n",
      "length                                        0.000\n",
      "annealing temperature                         0.000\n",
      "%cg_content                                   1.619\n",
      "specificity                                   0.000\n",
      "runs                                          0.000\n",
      "repeats                                      10.000\n",
      "\n",
      "\n",
      "===Reverse Primer=== attcgtgtagttgaaccctga\n",
      "Criterion                       Cost function score\n",
      "---------------------------------------------------\n",
      "length                                        0.000\n",
      "annealing temperature                         0.000\n",
      "%cg_content                                   0.000\n",
      "specificity                                   0.000\n",
      "runs                                          0.000\n",
      "repeats                                      10.000\n",
      "\n",
      "Total costs:                                 91.619\n",
      "\n",
      "Temperature Difference                       70.000\n",
      "\n",
      "\n",
      "Forward primer criteria is not met:\n",
      "-%cg_content\n",
      "-repeats\n",
      "\n",
      "Reverse primer criteria is not met:\n",
      "-repeats\n",
      "\n"
     ]
    }
   ],
   "source": [
    "class PrimerDesign(PrimerDesign):\n",
    "    \n",
    "    \n",
    "    def cost_objective_function_info(self, fp, rp):\n",
    "        txt = \"\"\n",
    "        txt+= \"===Forward Primer=== {}\\n\".format(fp)\n",
    "        txt += \"{0:<25} {1:>25}\\n\".format(\"Criterion\",\"Cost function score\")\n",
    "        txt += \"-\"*51 + \"\\n\"\n",
    "        txt += \"{0:<25} {1:>25.3f}\\n\".format(\"length\",self.cost_length(fp))\n",
    "        txt += \"{0:<25} {1:>25.3f}\\n\".format(\"annealing temperature\",self.cost_temperature(fp))\n",
    "        txt += \"{0:<25} {1:>25.3f}\\n\".format(\"%cg_content\",self.cost_cgcontent(fp))\n",
    "        txt += \"{0:<25} {1:>25.3f}\\n\".format(\"specificity\",self.cost_specificity(fp))\n",
    "        txt += \"{0:<25} {1:>25.3f}\\n\".format(\"runs\",self.cost_runs(fp))\n",
    "        txt += \"{0:<25} {1:>25.3f}\\n\\n\\n\".format(\"repeats\",self.cost_repeats(fp))\n",
    "        #the following codes change the complement of reverse primer from method self.func_select_random(self, sqtype='reverse', length = 20 ) into the reverse primer\n",
    "        #by replacing all the \"c\"s to \"g\"s and vice versa, and \"a\"s to \"t\"s and vice versa, then reversing the string to give the reverse primer in 5'- to -3' direction.\n",
    "        a = rp.replace(\"c\",\"g1\")\n",
    "        b = a.replace(\"g\",\"c\")\n",
    "        c = b.replace('a','t1')\n",
    "        d = c.replace('t','a')\n",
    "        e = d.replace('a1', 't')\n",
    "        f = e.replace('c1','g')\n",
    "        rpc = f[::-1]\n",
    "        txt+= \"===Reverse Primer=== {}\\n\".format(rpc)\n",
    "        txt += \"{0:<25} {1:>25}\\n\".format(\"Criterion\",\"Cost function score\")\n",
    "        txt += \"-\"*51 + \"\\n\"\n",
    "        txt += \"{0:<25} {1:>25.3f}\\n\".format(\"length\",self.cost_length(rp))\n",
    "        txt += \"{0:<25} {1:>25.3f}\\n\".format(\"annealing temperature\",self.cost_temperature(rp))\n",
    "        txt += \"{0:<25} {1:>25.3f}\\n\".format(\"%cg_content\",self.cost_cgcontent(rp))\n",
    "        txt += \"{0:<25} {1:>25.3f}\\n\".format(\"specificity\",self.cost_specificity(rp))\n",
    "        txt += \"{0:<25} {1:>25.3f}\\n\".format(\"runs\",self.cost_runs(rp))\n",
    "        txt += \"{0:<25} {1:>25.3f}\\n\".format(\"repeats\",self.cost_repeats(rp))\n",
    "        txt += \"\\n{0:<25} {1:>25.3f}\\n\".format(\"Total costs:\",self.cost_objective_function(fp, rp))\n",
    "        txt += \"\\n{0:<25} {1:>25.3f}\\n\".format(\"Temperature Difference\",self.cost_temperature_difference(fp,rp))\n",
    "        crit = ['length','annealing temperature','%cg_content','specificity','runs','repeats']\n",
    "        data = [self.cost_length(fp), self.cost_temperature(fp), self.cost_cgcontent(fp)  ,self.cost_specificity(fp),self.cost_runs(fp),self.cost_repeats(fp)] \n",
    "        data2 = [self.cost_length(rp), self.cost_temperature(rp), self.cost_cgcontent(rp)  ,self.cost_specificity(rp),self.cost_runs(rp),self.cost_repeats(rp)] \n",
    "        fpfail = []\n",
    "        rpfail = []\n",
    "        for i in range(len(data)):\n",
    "            if data[i] != 0 :\n",
    "                fpfail.append(crit[i])\n",
    "        for j in range(len(data)):\n",
    "            if data2[j] != 0:\n",
    "                rpfail.append(crit[j])\n",
    "        fpnotmet = ''\n",
    "        rpnotmet =''\n",
    "        for i in range(len(fpfail)):\n",
    "            fpnotmet += (\"-\" + fpfail[i] + \"\\n\")\n",
    "        for j in range(len(rpfail)):\n",
    "            rpnotmet += (\"-\" + rpfail[j] + \"\\n\")\n",
    "            \n",
    "        txt += \"\\n\\nForward primer criteria is not met:\\n{}\\nReverse primer criteria is not met:\\n{}\".format(fpnotmet,rpnotmet)\n",
    "        return txt\n",
    "\n",
    "#Test based on dna_sequence\n",
    "a = PrimerDesign(\"BIODW\")\n",
    "c = dna_sequence\n",
    "b = a.set_dna_sequence(c)\n",
    "fp = a.func_select_random(sqtype='forward', length =20)\n",
    "rp = a.func_select_random(sqtype='reverse', length =20)     \n",
    "print(a.cost_objective_function_info(fp,rp))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Task 10\n",
    "This method finds the forward primer and the reverse primer that will give an optimal cost objective function; all the criteria are met. This is done through generating possible solutions using random numbers to decide the primer pairs and updating the cost objective function associated such that it will find the primer pairs that corresponds to an optimal value of 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "class PrimerDesign(PrimerDesign): \n",
    "    \n",
    "    def func_simulated_annealing(self):\n",
    "\n",
    "        def get_primer_neighbour(sq):\n",
    "            length = len(sq)\n",
    "            oldstartposition = b.find(sq)\n",
    "            newlength = random.randint(self.min_length, self.max_length + 1)\n",
    "            if oldstartposition >= (self.fp_end - length):\n",
    "                newstart = oldstartposition + random.randint(-4,-1)\n",
    "                newend = newstart + newlength\n",
    "                newfp = self.dna_sequence[newstart:newend+1]\n",
    "            elif oldstartposition == (self.fp_start):\n",
    "                newstart= oldstartposition + random.randint(1,4)\n",
    "                newend = newstart + newlength\n",
    "                newfp = self.dna_sequence[newstart:newend+1]\n",
    "            else:\n",
    "                newstart= oldstartposition + random.randint(-4,4)\n",
    "                newend = newstart + newlength\n",
    "                newfp = self.dna_sequence[newstart:newend+1]\n",
    "            return newfp\n",
    "\n",
    "        def get_primer_neighbour2(sq):\n",
    "            length = len(sq)\n",
    "            oldstartposition = b.find(sq)\n",
    "            newlength = random.randint(self.min_length, self.max_length + 1)\n",
    "            if oldstartposition >= (self.rp_end - length):\n",
    "                newstart = oldstartposition + random.randint(-4,-1)\n",
    "                newend = newstart + newlength\n",
    "                newrp = self.dna_sequence[newstart:newend+1]\n",
    "            elif oldstartposition == (self.rp_start):\n",
    "                newstart= oldstartposition + random.randint(1,4)\n",
    "                newend = newstart + newlength\n",
    "                newrp = self.dna_sequence[newstart:newend+1]\n",
    "            else:\n",
    "                newstart= oldstartposition + random.randint(-4,4)\n",
    "                newend = newstart + newlength\n",
    "                newrp = self.dna_sequence[newstart:newend+1]\n",
    "            return newrp\n",
    "\n",
    "\n",
    "        temperature = self.initial_temperature\n",
    "        stopping_temperature = self.stopping_temperature\n",
    "        drop = self.drop_fraction      \n",
    "        fp_current = a.func_select_random(sqtype='forward', length =20)\n",
    "        rp_current = a.func_select_random(sqtype='reverse', length =20)\n",
    "        cost = self.cost_objective_function(fp_current, rp_current)\n",
    "        current_cost = cost\n",
    "\n",
    "        while temperature>stopping_temperature:\n",
    "            fp_new = get_primer_neighbour(fp_current)\n",
    "            rp_new = get_primer_neighbour2(rp_current)\n",
    "            new_cost = self.cost_objective_function(fp_new, rp_new)\n",
    "            if( new_cost < current_cost ):\n",
    "                fp_current = fp_new\n",
    "                rp_current = rp_new\n",
    "                current_cost = new_cost \n",
    "            else:\n",
    "                delta = new_cost - current_cost\n",
    "                acceptance_probability = math.exp(-delta/temperature) \n",
    "                num = random.uniform(0.0,1.0)\n",
    "                if ( acceptance_probability > num):\n",
    "                    fp_current = fp_new\n",
    "                    rp_current = rp_new\n",
    "                    current_cost = new_cost \n",
    "                else:\n",
    "                    pass\n",
    "            temperature*=drop\n",
    "        #the method returns the optimal forward primer, reverse primer and the cost objective function\n",
    "        return fp_current, rp_current, current_cost\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Store the DNA sequence given to you in the variable below \n",
    "We used allele 1 of Homo Sapiens APC gene regulator of WNT signaling pathway (from Biology worksheet)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [],
   "source": [
    "dna_sequence = '''1 gtattggtgc agcccgccag ggtgtcactg gagacagaat ggaggtgctg ccggactcgg\n",
    "61 aaatggggac acatgcattg gtggccaaaa gagagaggag acaaaaccgc tgcagatggc\n",
    "121 tgatgtgaat ctagtggaaa gagctactgg ggatgagaga aagaggagga ggcaggaagt\n",
    "181 acttaaacaa ctacaaggaa gtattgaaga tgaagctatg gcttcttctg gacagattga\n",
    "241 tttattagag cgtcttaaag agcttaactt agatagcagt aatttccctg gagtaaaact\n",
    "301 gcggtcaaaa atgtccctcc gttcttatgg aagccgggaa ggatctgtat caagccgttc\n",
    "361 tggagagtgc agtcctgttc ctatgggttc atttccaaga agagggtttg taaatggaag\n",
    "421 cagagaaagt actggatatt tagaagaact tgagaaagag aggtcattgc ttcttgctga\n",
    "481 tcttgacaaa gaagaaaagg aaaaagactg gtattacgct caacttcaga atctcactaa\n",
    "541 aagaatagat agtcttcctt taactgaaaa tttttcctta caaacagata tgaccagaag\n",
    "601 gcaattggaa tatgaagcaa ggcaaatcag agttgcgatg gaagaacaac taggtacctg\n",
    "661 ccaggatatg gaaaaacgag cacagcgaag aatagccaga attcagcaaa tcgaaaagga\n",
    "721 catacttcgt atacgacagc ttttacagtc ccaagcaaca gaagcagaga ggtcatctca\n",
    "781 gaacaagcat gaaaccggct cacatgatgc tgagcggcag aatgaaggtc aaggagtggg\n",
    "841 agaaatcaac atggcaactt ctggtaatgg tcagggttca actacacgaa tggaccatga\n",
    "901 aacagccagt gttttgagtt ctagtagcac acactctgca cctcgaaggc tgacaagtca\n",
    "961 tctgggaacc aagatacgcg cttactgtga aacctgttgg gagtggcagg aagctcatga\n",
    "1021 accaggcatg gaccaggaca aaaatccaat gccagctcct gttgaacatc agatctgtcc\n",
    "1081 tgctgtgtgt gttctaatga aactttcatt tgatgaagag catagacatg caatgaatga\n",
    "1141 actaggggga ctacaggcca ttgcagaatt attgcaagtg gactgtgaaa tgtatgggct\n",
    "1201 tactaatgac cactacagta ttacactaag acgatatgct ggaatggctt tgacaaactt\n",
    "1261 gacttttgga gatgtagcca acaaggctac gctatgctct atgaaaggct gcatgagagc\n",
    "1321 acttgtggcc caactaaaat ctgaaagtga agacttacag caggttattg cgagtgtttt\n",
    "1381 gaggaatttg tcttggcgag cagatgtaaa tagtaaaaag acgttgcgag aagttggaag\n",
    "1441 tgtgaaagca ttgatggaat gtgctttaga agttaaaaag gaatcaaccc tcaaaagcgt\n",
    "1501 attgagtgcc ttatggaatt tgtcagcaca ttgcactgag aataaagctg atatatgtgc\n",
    "1561 tgtagatggt gcacttgcat ttttggttgg cactcttact taccggagcc agacaaacac\n",
    "1621 tttagccatt attgaaagtg gaggtgggat attacggaat gtgtccagct tgatagctac\n",
    "1681 aaatgaggac cacaggcaaa tcctaagaga gaacaactgt ctacaaactt tattacaaca\n",
    "1741 cttaaaatct catagtttga caatagtcag taatgcatgt ggaactttgt ggaatctctc\n",
    "1801 agcaagaaat cctaaagacc aggaagcatt atgggacatg ggggcagtta gcatgctcaa\n",
    "1861 gaacctcatt cattcaaagc acaaaatgat tgctatggga agtgctgcag ctttaaggaa\n",
    "1921 tctcatggca aataggcctg cgaagtacaa ggatgccaat attatgtctc ctggctcaag\n",
    "1981 cttgccatct cttcatgtta ggaaacaaaa agccctagaa gcagaattag atgctcagca\n",
    "2041 cttatcagaa acttttgaca atatagacaa tttaagtccc aaggcatctc atcgtagtaa\n",
    "2101 gcagagacac aagcaaagtc tctatggtga ttatgttttt gacaccaatc gacatgatga\n",
    "2161 taataggtca gacaatttta atactggcaa catgactgtc ctttcaccat atttgaatac\n",
    "2221 tacagtgtta cccagctcct cttcatcaag aggaagctta gatagttctc gttctgaaaa\n",
    "2281 agatagaagt ttggagagag aacgcggaat tggtctaggc aactaccatc cagcaacaga\n",
    "2341 aaatccagga acttcttcaa agcgaggttt gcagatctcc accactgcag cccagattgc\n",
    "2401 caaagtcatg gaagaagtgt cagccattca tacctctcag gaagacagaa gttctgggtc\n",
    "2461 taccactgaa ttacattgtg tgacagatga gagaaatgca cttagaagaa gctctgctgc\n",
    "2521 ccatacacat tcaaacactt acaatttcac taagtcggaa aattcaaata ggacatgttc\n",
    "2581 tatgccttat gccaaattag aatacaagag atcttcaaat gatagtttaa atagtgtcag\n",
    "2641 tagtagtgat ggttatggta aaagaggtca aatgaaaccc tcgattgaat cctattctga\n",
    "2701 agatgatgaa agtaagtttt gcagttatgg tcaataccca gccgacctag cccataaaat\n",
    "2761 acatagtgca aatcatatgg atgataatga tggagaacta gatacaccaa taaattatag\n",
    "2821 tcttaaatat tcagatgagc agttgaactc tggaaggcaa agtccttcac agaatgaaag\n",
    "2881 atgggcaaga cccaaacaca taatagaaga tgaaataaaa caaagtgagc aaagacaatc\n",
    "2941 aaggaatcaa agtacaactt atcctgttta tactgagagc actgatgata aacacctcaa\n",
    "3001 gttccaacca cattttggac agcaggaatg tgtttctcca tacaggtcac ggggagccaa\n",
    "3061 tggttcagaa acaaatcgag tgggttctaa tcatggaatt aatcaaaatg taagccagtc\n",
    "3121 tttgtgtcaa gaagatgact atgaagatga taagcctacc aattatagtg aacgttactc\n",
    "3181 tgaagaagaa cagcatgaag aagaagagag accaacaaat tatagcataa aatataatga\n",
    "3241 agagaaacgt catgtggatc agcctattga ttatagttta aaatatgcca cagatattcc\n",
    "3301 ttcatcacag aaacagtcat tttcattctc aaagagttca tctggacaaa gcagtaaaac\n",
    "3361 cgaacatatg tcttcaagca gtgagaatac gtccacacct tcatctaatg ccaagaggca\n",
    "3421 gaatcagctc catccaagtt ctgcacagag tagaagtggt cagcctcaaa aggctgccac\n",
    "3481 ttgcaaagtt tcttctatta accaagaaac aatacagact tattgtgtag aagatactcc\n",
    "3541 aatatgtttt tcaagatgta gttcattatc atctttgtca tcagctgaag atgaaatagg\n",
    "3601 atgtaatcag acgacacagg aagcagattc tgctaatacc ctgcaaatag cagaaataaa\n",
    "3661 agaaaagatt ggaactaggt cagctgaaga tcctgtgagc gaagttccag cagtgtcaca\n",
    "3721 gcaccctaga accaaatcca gcagactgca gggttctagt ttatcttcag aatcagccag\n",
    "3781 gcacaaagct gttgaatttt cttcaggagc gaaatctccc tccaaaagtg gtgctcagac\n",
    "3841 acccaaaagt ccacctgaac actatgttca ggagacccca ctcatgttta gcagatgtac\n",
    "3901 ttctgtcagt tcacttgata gttttgagag tcgttcgatt gccagctccg ttcagagtga\n",
    "3961 accatgcagt ggaatggtaa gtggcattat aagccccagt gatcttccag atagccctgg\n",
    "4021 acaaaccatg ccaccaagca gaagtaaaac acctccacca cctcctcaaa cagctcaaac\n",
    "4081 caagcgagaa gtacctaaaa ataaagcacc tactgctgaa aagagagaga gtggacctaa\n",
    "4141 gcaagctgca gtaaatgctg cagttcagag ggtccaggtt cttccagatg ctgatacttt\n",
    "4201 attacatttt gccacggaaa gtactccaga tggattttct tgttcatcca gcctgagtgc\n",
    "4261 tctgagcctc gatgagccat ttatacagaa agatgtggaa ttaagaataa tgcctccagt\n",
    "4321 tcaggaaaat gacaatggga atgaaacaga atcagagcag cctaaagaat caaatgaaaa\n",
    "4381 ccaagagaaa gaggcagaaa aaactattga ttctgaaaag gacctattag atgattcaga\n",
    "4441 tgatgatgat attgaaatac tagaagaatg tattatttct gccatgccaa caaagtcatc\n",
    "4501 acgtaaagca aaaaagccag cccagactgc ttcaaaatta cctccacctg tggcaaggaa\n",
    "4561 accaagtcag ctgcctgtgt acaaacttct accatcacaa aacaggttgc aaccccaaaa\n",
    "4621 gcatgttagt tttacaccgg gggatgatat gccacgggtg tattgtgttg aagggacacc\n",
    "4681 tataaacttt tccacagcta catctctaag tgatctaaca atcgaatccc ctccaaatga\n",
    "4741 gttagctgct ggagaaggag ttagaggagg ggcacagtca ggtgaatttg aaaaacgaga\n",
    "4801 taccattcct acagaaggca gaagtacaga tgaggctcaa ggaggaaaaa cctcatctgt\n",
    "4861 aaccatacct gaattggatg acaataaagc agaggaaggt gatattcttg cagaatgcat\n",
    "4921 taattctgct atgcccaaag ggaaaagtca caagcctttc cgtgtgaaaa agataatgga\n",
    "4981 ccaggtccag caagcatctg cgtcttcttc tgcacccaac aaaaatcagt tagatggtaa\n",
    "5041 gaaaaagaaa ccaacttcac cagtaaaacc tataccacaa aatactgaat ataggacacg\n",
    "5101 tgtaagaaaa aatgcagact caaaaaataa tttaaatgct gagagagttt tctcagacaa\n",
    "5161 caaagattca aagaaacaga atttgaaaaa taattccaag gtcttcaatg ataagctccc\n",
    "5221 aaataatgaa gatagagtca gaggaagttt tgcttttgat tcacctcatc attacacgcc\n",
    "5281 tattgaagga actccttact gtttttcacg aaatgattct ttgagttctc tagattttga\n",
    "5341 tgatgatgat gttgaccttt ccagggaaaa ggctgaatta agaaaggcaa aagaaaataa\n",
    "5401 ggaatcagag gctaaagtta ccagccacac agaactaacc tccaaccaac aatcagctaa\n",
    "5461 taagacacaa gctattgcaa agcagccaat aaatcgaggt cagcctaaac ccatacttca\n",
    "5521 gaaacaatcc acttttcccc agtcatccaa agacatacca gacagagggg cagcaactga\n",
    "5581 tgaaaagtta cagaattttg ctattgaaaa tactccggtt tgcttttctc ataattcctc\n",
    "5641 tctgagttct ctcagtgaca ttgaccaaga aaacaacaat aaagaaaatg aacctatcaa\n",
    "5701 agagactgag ccccctgact cacagggaga accaagtaaa cctcaagcat caggctatgc\n",
    "5761 tcctaaatca tttcatgttg aagatacccc agtttgtttc tcaagaaaca gttctctcag\n",
    "5821 ttctcttagt attgactctg aagatgacct gttgcaggaa tgtataagct ccgcaatgcc\n",
    "5881 aaaaaagaaa aagccttcaa gactcaaggg tgataatgaa aaacatagtc ccagaaatat\n",
    "5941 gggtggcata ttaggtgaag atctgacact tgatttgaaa gatatacaga gaccagattc\n",
    "6001 agaacatggt ctatcccctg attcagaaaa ttttgattgg aaagctattc aggaaggtgc\n",
    "6061 aaattccata gtaagtagtt tacatcaagc tgctgctgct gcatgtttat ctagacaagc\n",
    "6121 ttcgtctgat tcagattcca tcctttccct gaaatcagga atctctctgg gatcaccatt\n",
    "6181 tcatcttaca cctgatcaag aagaaaaacc ctttacaagt aataaaggcc cacgaattct\n",
    "6241 aaaaccaggg gagaaaagta cattggaaac taaaaagata gaatctgaaa gtaaaggaat\n",
    "6301 caaaggagga aaaaaagttt ataaaagttt gattactgga aaagttcgat ctaattcaga\n",
    "6361 aatttcaggc caaatgaaac agccccttca agcaaacatg ccttcaatct ctcgaggcag\n",
    "6421 gacaatgatt catattccag gagttcgaaa tagctcctca agtacaagtc ctgtttctaa\n",
    "6481 aaaaggccca ccccttaaga ctccagcctc caaaagccct agtgaaggtc aaacagccac\n",
    "6541 cacttctcct agaggagcca agccatctgt gaaatcagaa ttaagccctg ttgccaggca\n",
    "6601 gacatcccaa ataggtgggt caagtaaagc accttctaga tcaggatcta gagattcgac\n",
    "6661 cccttcaaga cctgcccagc aaccattaag tagacctata cagtctcctg gccgaaactc\n",
    "6721 aatttcccct ggtagaaatg gaataagtcc tcctaacaaa ttatctcaac ttccaaggac\n",
    "6781 atcatcccct agtactgctt caactaagtc ctcaggttct ggaaaaatgt catatacatc\n",
    "6841 tccaggtaga cagatgagcc aacagaacct taccaaacaa acaggtttat ccaagaatgc\n",
    "6901 cagtagtatt ccaagaagtg agtctgcctc caaaggacta aatcagatga ataatggtaa\n",
    "6961 tggagccaat aaaaaggtag aactttctag aatgtcttca actaaatcaa gtggaagtga\n",
    "7021 atctgataga tcagaaagac ctgtattagt acgccagtca actttcatca aagaagctcc\n",
    "7081 aagcccaacc ttaagaagaa aattggagga atctgcttca tttgaatctc tttctccatc\n",
    "7141 atctagacca gcttctccca ctaggtccca ggcacaaact ccagttttaa gtccttccct\n",
    "7201 tcctgatatg tctctatcca cacattcgtc tgttcaggct ggtggatggc gaaaactccc\n",
    "7261 acctaatctc agtcccacta tagagtataa tgatggaaga ccagcaaagc gccatgatat\n",
    "7321 tgcacggtct cattctgaaa gtccttctag acttccaatc aataggtcag gaacctggaa\n",
    "7381 acgtgagcac agcaaacatt catcatccct tcctcgagta agcacttgga gaagaactgg\n",
    "7441 aagttcatct tcaattcttt ctgcttcatc agaatccagt gaaaaagcaa aaagtgagga\n",
    "7501 tgaaaaacat gtgaactcta tttcaggaac caaacaaagt aaagaaaacc aagtatccgc\n",
    "7561 aaaaggaaca tggagaaaaa taaaagaaaa tgaattttct cccacaaata gtacttctca\n",
    "7621 gaccgtttcc tcaggtgcta caaatggtgc tgaatcaaag actctaattt atcaaatggc\n",
    "7681 acctgctgtt tctaaaacag aggatgtttg ggtgagaatt gaggactgtc ccattaacaa\n",
    "7741 tcctagatct ggaagatctc ccacaggtaa tactcccccg gtgattgaca gtgtttcaga\n",
    "7801 aaaggcaaat ccaaacatta aagattcaaa agataatcag gcaaaacaaa atgtgggtaa\n",
    "7861 tggcagtgtt cccatgcgta ccgtgggttt ggaaaatcgc ctgaactcct ttattcaggt\n",
    "7921 ggatgcccct gaccaaaaag gaactgagat aaaaccagga caaaataatc ctgtccctgt\n",
    "7981 atcagagact aatgaaagtt ctatagtgga acgtacccca ttcagttcta gcagctcaag\n",
    "8041 caaacacagt tcacctagtg ggactgttgc tgccagagtg actcctttta attacaaccc\n",
    "8101 aagccctagg aaaagcagcg cagatagcac ttcagctcgg ccatctcaga tcccaactcc\n",
    "8161 agtgaataac aacacaaaga agcgagattc caaaactgac agcacagaat ccagtggaac\n",
    "8221 ccaaagtcct aagcgccatt ctgggtctta ccttgtgaca tctgtttaaa agagaggaag\n",
    "8281 aatgaaacta agaaaattct atgttaatta caactgctat atagacattt tgtttcaaat\n",
    "8341 gaaactttaa aagactgaaa aattttgtaa ataggtttga ttcttgttag agggtttttg\n",
    "8401 ttctggaagc catatttgat agtatacttt gtcttcactg gtcttatttt gggaggcact\n",
    "8461 cttgatggtt aggaaaaaaa tagtaaagcc aagtatgttt gtacagtatg ttttacatgt\n",
    "8521 atttaaagta gcatcccatc ccaacttcct ttaattattg cttgtcttaa aataatgaac\n",
    "8581 actacagata gaaaatatga tatattgctg ttatcaatca tttctagatt ataaactgac\n",
    "8641 taaacttaca tcagggaaaa attggtattt atgcaaaaaa aaatgttttt gtccttgtga\n",
    "8701 gtccatctaa catcataatt aatcatgtgg ctgtgaaatt cacagtaata tggttcccga\n",
    "8761 tgaacaagtt tacccagcct gctttgcttt actgcatgaa tgaaactgat ggttcaattt\n",
    "8821 cagaagtaat gattaacagt tatgtggtca catgatgtgc atagagatag ctacagtgta\n",
    "8881 ataatttaca ctattttgtg ctccaaacaa aacaaaaatc tgtgtaactg taaaacattg\n",
    "8941 aatgaaacta ttttacctga actagatttt atctgaaagt aggtagaatt tttgctatgc\n",
    "9001 tgtaatttgt tgtatattct ggtatttgag gtgagatggc tgctctttta ttaatgagac\n",
    "9061 atgaattgtg tctcaacaga aactaaatga acatttcaga ataaattatt gctgtatgta\n",
    "9121 aactgttact gaaattggta tttgtttgaa gggtcttgtt tcacatttgt attaataatt\n",
    "9181 gtttaaaatg cctcttttaa aagcttatat aaattttttt cttcagcttc tatgcattaa\n",
    "9241 gagtaaaatt cctcttactg taataaaaac aattgaagaa gactgttgcc acttaaccat\n",
    "9301 tccatgcgtt ggcacttatc tattcctgaa atttctttta tgtgattagc tcatcttgat\n",
    "9361 ttttaatatt tttccactta aacttttttt tcttactcca ctggagctca gtaaaagtaa\n",
    "9421 attcatgtaa tagcaatgca agcagcctag cacagactaa gcattgagca taataggccc\n",
    "9481 acataatttc ctctttctta atattataga attctgtact tgaaattgat tcttagacat\n",
    "9541 tgcagtctct tcgaggcttt acagtgtaaa ctgtcttgcc ccttcatctt cttgttgcaa\n",
    "9601 ctgggtctga catgaacact ttttatcacc ctgtatgtta gggcaagatc tcagcagtga\n",
    "9661 agtataatca gcactttgcc atgctcagaa aattcaaatc acatggaact ttagaggtag\n",
    "9721 atttaatacg attaagatat tcagaagtat attttagaat ccctgcctgt taaggaaact\n",
    "9781 ttatttgtgg taggtacagt tctggggtac atgttaagtg tccccttata cagtggaggg\n",
    "9841 aagtcttcct tcctgaagga aaataaactg acacttatta actaagataa tttacttaat\n",
    "9901 atatcttccc tgatttgttt taaaagatca gagggtgact gatgatacat gcatacatat\n",
    "9961 ttgttgaata aatgaaaatt tatttttagt gataagattc atacactctg tatttgggga\n",
    " '''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Instantiate your class and read in the DNA sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = PrimerDesign(\"BIODW\")\n",
    "dna = dna_sequence\n",
    "b = a.set_dna_sequence(dna)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### If you need to adjust any parameter from their default values in the init method, do it here"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [],
   "source": [
    "#all default values stay"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Show the outcome of your testing and the functions in the subsequent cells "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "===Forward Primer=== gtcctgttcctatgggttc\n",
      "Criterion                       Cost function score\n",
      "---------------------------------------------------\n",
      "length                                        0.000\n",
      "annealing temperature                         0.000\n",
      "%cg_content                                   0.000\n",
      "specificity                                   0.000\n",
      "runs                                          0.000\n",
      "repeats                                       0.000\n",
      "\n",
      "\n",
      "===Reverse Primer=== gttgttcttccatcgcaact\n",
      "Criterion                       Cost function score\n",
      "---------------------------------------------------\n",
      "length                                        0.000\n",
      "annealing temperature                         0.000\n",
      "%cg_content                                   0.000\n",
      "specificity                                   0.000\n",
      "runs                                          0.000\n",
      "repeats                                       0.000\n",
      "\n",
      "Total costs:                                  0.000\n",
      "\n",
      "Temperature Difference                        0.000\n",
      "\n",
      "\n",
      "Forward primer criteria is not met:\n",
      "\n",
      "Reverse primer criteria is not met:\n",
      "\n"
     ]
    }
   ],
   "source": [
    "fp,rp,x=a.func_simulated_annealing()      \n",
    "print(a.cost_objective_function_info(fp,rp))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid character in identifier (<ipython-input-215-524a050aa62c>, line 3)",
     "output_type": "error",
     "traceback": [
      "\u001b[1;36m  File \u001b[1;32m\"<ipython-input-215-524a050aa62c>\"\u001b[1;36m, line \u001b[1;32m3\u001b[0m\n\u001b[1;33m     What were the simulation parameters that you used (in Step 1.1 of the\u001b[0m\n\u001b[1;37m    ^\u001b[0m\n\u001b[1;31mSyntaxError\u001b[0m\u001b[1;31m:\u001b[0m invalid character in identifier\n"
     ]
    }
   ],
   "source": [
    "QUESTIONS\n",
    "\n",
    " What were the simulation parameters that you used (in Step 1.1 of the\n",
    "algorithm)?\n",
    "\n",
    "I set the initial temperature to be the maximum annealing temperature, which is (0.6*22*4) + (0.4*22*2) = 70.4 degrees Celcius.\n",
    "The stopping temperature is set to 0.01 so as to ensure that the method runs sufficient tests to generate the best primer pairs for the optimal cost objective function.\n",
    "Decreasing factor is set as 0.999 to create the slightest increment in temperature so as to generate sufficient tests for optimal solution.\n",
    "\n",
    " From the output of your code, how well does the final solution meet all\n",
    "the criteria?\n",
    "\n",
    "The output of my code generates primer pairs that have 0 value for all the criteria cost functions, indicating the optimal cost objective function. \n",
    "Hence, they meet the criteria and are suitable for PCR.\n",
    "\n",
    " How many iterations were required to obtain this solution?\n",
    "8854 iterations were required. ( number of iterations = (ln(0.01/70.4))/ln(0.999) )\n",
    "\n",
    " Will a brute-force algorithm (one that considers every single possible\n",
    "solution) require fewer or more iterations? Explain.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
